File: src/main/java/fr/lewon/web/bot/util/HtmlAnalyzer.java
package fr.lewon.web.bot.util;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import fr.lewon.bot.http.util.JsonHelper;
import fr.lewon.web.bot.entities.activities.Mission;
import fr.lewon.web.bot.entities.battle.BattleMob;
import fr.lewon.web.bot.entities.battle.BattlePlayer;
import fr.lewon.web.bot.entities.battle.TowerOfFameOpponentPremise;
import fr.lewon.web.bot.entities.champions.ChampionPremise;
import fr.lewon.web.bot.entities.girls.Girl;
import fr.lewon.web.bot.entities.quests.QuestStep;
import fr.lewon.web.bot.entities.response.ChampionData;
import fr.lewon.web.bot.entities.response.UserInfos;
import fr.lewon.web.bot.entities.shop.Item;
import fr.lewon.web.bot.entities.shop.Shop;

public enum HtmlAnalyzer {

	INSTANCE;

	private HtmlAnalyzer() {}

	private JsonHelper jsonHelper = new JsonHelper();

	public List<Integer> getCompetitions(String activityPage) throws IOException {
		String regex = "<div class=\"contest\" id_contest=\"([0-9]+)\">.*?<div class=\"(.*?)\"";
		Matcher matcher = matchPattern(activityPage, regex);
		List<Integer> competitions = new ArrayList<>();
		while (matcher.find()) {
			String divEndedClass = matcher.group(2);
			if ("contest_header ended".equals(divEndedClass)) {
				competitions.add(Integer.parseInt(matcher.group(1)));
			}
		}
		return competitions;
	}

	public ChampionData getChampionData(String championContent) throws IOException {
		String regex = "var championData = (\\{.*?});";
		Matcher matcher = matchPattern(championContent, regex);
		if (matcher.find()) {
			return jsonHelper.jsonToObject(ChampionData.class, matcher.group(1));
		}
		return null;
	}

	public List<ChampionPremise> getChampionsIds(String championsMapContent) {
		String regex = "<a href=\"champions/([0-9]+)\" class=\"champion-lair\""
				+ ".*?"
				+ "<div class=\"champion-lair-name map-label-link\">"
				+ "(.*?)"
				+ "</div>";
		Matcher matcher = matchPattern(championsMapContent, regex);
		List<ChampionPremise> premises = new ArrayList<>();
		while (matcher.find()) {
			int id = Integer.parseInt(matcher.group(1));
			int waitTime = getWaitTime(matcher.group(2));
			premises.add(new ChampionPremise(id, waitTime));
		}
		return premises;
	}

	private int getWaitTime(String championLairContent) {
		String regex = "<div rel=\"timer\" timer=\"([0-9]+)\">";
		Matcher matcher = matchPattern(championLairContent, regex);
		if (matcher.find()) {
			Long timer = Long.parseLong(matcher.group(1));
			Integer waitTime = (int) (timer - System.currentTimeMillis() / 1000);
			return waitTime < 0 ? 0 : waitTime;
		}
		return 0;
	}


	public QuestStep[] getQuestSteps(String questContent) throws IOException {
		String regex = "Q.steps = (\\[.*?\\]);";
		Matcher matcher = matchPattern(questContent, regex);
		if (!matcher.find()) {
			return null;
		}
		String json = matcher.group(1).replace("\"cost\":[]", "\"cost\":{}");
		return jsonHelper.jsonToObject(QuestStep[].class, json);
	}

	public List<Girl> findAllGirls(String haremContent) throws IOException {
		List<Girl> girls = new ArrayList<>();
		String regex = "girlsDataList\\['[0-9]+'\\] = (\\{.*?};)";
		Matcher matcher = matchPattern(haremContent, regex);

		while (matcher.find()) {
			Girl girl = jsonHelper.jsonToObject(Girl.class, matcher.group(1));
			girls.add(girl);
		}
		return girls;
	}

	public UserInfos getPlayerInfos(String homeContent) throws IOException {
		String regex = "Hero.infos = (\\{.*?};)";
		Matcher matcher = matchPattern(homeContent, regex);

		if (!matcher.find()) {
			return null;
		}

		return jsonHelper.jsonToObject(UserInfos.class, matcher.group(1));
	}

	public BattlePlayer findOpponentBattlePlayer(String content) throws IOException {
		String regex = "hh_battle_players =.*?\\{.*?},.*?(\\{.*?})";
		Matcher matcher = matchPattern(content, regex);

		if (!matcher.find()) {
			return null;
		}
		String battlePlayerStr = matcher.group(1);
		return jsonHelper.jsonToObject(BattlePlayer.class, battlePlayerStr);
	}

	public BattleMob findOpponentBattleMob(String battleTrollContent) throws IOException {
		String regex = "hh_battle_players =.*?\\{.*?},.*?(\\{.*?})";
		Matcher matcher = matchPattern(battleTrollContent, regex);

		if (!matcher.find()) {
			return null;
		}
		String battlePlayerStr = matcher.group(1);
		return jsonHelper.jsonToObject(BattleMob.class, battlePlayerStr);
	}

	public String getCurrentWorldId(String mapContent) {
		String regex = "<a class=\"link-world\" href=\"/world/([0-9])\"";
		Matcher matcher = matchPattern(mapContent, regex);
		String lastWorldId = null;
		while (matcher.find()) {
			lastWorldId = matcher.group(1);
		}
		return lastWorldId;
	}

	public String getTrollId(String worldContent) {
		String regex = "id_troll=([0-9])";
		Matcher matcher = matchPattern(worldContent, regex);
		if (matcher.find()) {
			return matcher.group(1);
		}
		return null;
	}

	public Integer getNextStock(String shopContent) {
		String regex = "<span rel=\"count\" time=\"([0-9]+)\">";
		Matcher matcher = matchPattern(shopContent, regex);
		if (matcher.find()) {
			return Integer.parseInt(matcher.group(1));
		}
		return null;
	}

	public Shop getShop(String shopContent) throws IOException {
		String regex = "<div tab class=\"armor\" type=\"armor\">(.*?)<button rel=\"buy\"";
		Matcher matcher = matchPattern(shopContent, regex);
		if (matcher.find()) {
			String subShopContent = matcher.group(1)
					.replace("&quot;", "\"")
					.replace("data-d=", "\ndata-d=");
			List<Item> books = getBooks(subShopContent);
			List<Item> gifts = getGifts(subShopContent);
			return new Shop(books, gifts);
		}
		return null;
	}

	private List<Item> getGifts(String subShopContent) throws IOException {
		String regex = "data-d=\"(\\{.*?\"type\":\"gift\".*?})\">";
		Matcher matcher = matchPattern(subShopContent, regex, false);
		List<Item> gifts = new ArrayList<>();
		while (matcher.find()) {
			gifts.add(jsonHelper.jsonToObject(Item.class, matcher.group(1)));
		}
		return gifts;
	}

	private List<Item> getBooks(String subShopContent) throws IOException {
		String regex = "data-d=\"(\\{.*?\"type\":\"potion\".*?})\">";
		Matcher matcher = matchPattern(subShopContent, regex, false);
		List<Item> books = new ArrayList<>();
		while (matcher.find()) {
			books.add(jsonHelper.jsonToObject(Item.class, matcher.group(1)));
		}
		return books;
	}

	public List<TowerOfFameOpponentPremise> findHallOfFameOpponents(String content) {
		String regexLeadTable = "<div class=\"lead_table_view\">.*?<tbody class=\"leadTable\" sorting_table>(.*?)<\\/div>";
		Matcher matcherLeadTable = matchPattern(content, regexLeadTable);
		if (!matcherLeadTable.find()) {
			return new ArrayList<>();
		}
		content = matcherLeadTable.group(1);

		String regex = "<tr sorting_id=\"(.*?)\".*?<span class=\"nickname\">(.*?)<\\/span>.*?<td>.*?([0-9]+).*?<td>([0-9])\\/3.*?<\\/tr>";
		Matcher matcher = matchPattern(content, regex);

		List<TowerOfFameOpponentPremise> premises = new ArrayList<>();
		while (matcher.find()) {
			int timesFought = Integer.parseInt(matcher.group(4));
			if (timesFought == 3) {
				continue;
			}
			String id = matcher.group(1);
			String nickName = matcher.group(2);
			int lvl = Integer.parseInt(matcher.group(3));
			premises.add(new TowerOfFameOpponentPremise(id, nickName, lvl));
		}

		return premises;
	}

	private Matcher matchPattern(String content, String regex) {
		return matchPattern(content, regex, true);
	}

	private Matcher matchPattern(String content, String regex, boolean flatten) {
		content = flatten ? flattenContent(content) : content;
		Pattern pattern = Pattern.compile(regex);
		return pattern.matcher(content);
	}

	private String flattenContent(String content) {
		return content.replaceAll("[\r\n]+", " ");
	}

	public List<Mission> getMissions(String activityPage) throws IOException {
		List<Mission> missions = new ArrayList<>();
		String regexMissionsWrap = "<div class=\"missions_wrap\">(.*?)<script type=\"text/javascript\"";
		Matcher matcherMissionsWrap = matchPattern(activityPage, regexMissionsWrap);
		if (!matcherMissionsWrap.find()) {
			return missions;
		}
		String missionsWrapContent = matcherMissionsWrap.group(1);
		String missionBodyRegex = "<div class=\"mission_object sub_block (.*?)\" data-d='(\\{.*?})'.*?<button rel=\"mission_start\" class=\"blue_text_button\" (.*?)>";
		Matcher matcherMissions = matchPattern(missionsWrapContent, missionBodyRegex);
		while (matcherMissions.find()) {
			String rarity = matcherMissions.group(1);
			String missionBody = matcherMissions.group(2);
			String displayStart = matcherMissions.group(3);
			boolean startable = displayStart == null || "".equals(displayStart);

			Mission mission = jsonHelper.jsonToObject(Mission.class, missionBody);
			mission.setRarity(rarity);
			mission.setStartable(startable);
			missions.add(mission);
		}
		return missions;
	}


}

File: src/main/kotlin/fr/lewon/bot/hh/rest/HtmlAnalyzer.kt
package fr.lewon.bot.hh.rest

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import fr.lewon.bot.hh.entities.activities.Mission
import fr.lewon.bot.hh.entities.activities.PlaceOfPower
import fr.lewon.bot.hh.entities.battle.BattleMob
import fr.lewon.bot.hh.entities.battle.BattlePlayer
import fr.lewon.bot.hh.entities.battle.TowerOfFameOpponentPremise
import fr.lewon.bot.hh.entities.champions.ChampionPremise
import fr.lewon.bot.hh.entities.event.EventData
import fr.lewon.bot.hh.entities.girls.Girl
import fr.lewon.bot.hh.entities.quests.QuestStep
import fr.lewon.bot.hh.entities.response.ChampionData
import fr.lewon.bot.hh.entities.response.UserInfos
import fr.lewon.bot.hh.entities.shop.Item
import fr.lewon.bot.hh.entities.shop.Shop
import java.io.IOException
import java.util.*
import java.util.regex.Matcher
import java.util.regex.Pattern

enum class HtmlAnalyzer {
    INSTANCE;

    private val objectMapper = ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    @Throws(IOException::class)
    fun getCompetitions(activityPage: String): List<Int> {
        val regex = "<div class=\"contest\" id_contest=\"([0-9]+)\">.*?<div class=\"(.*?)\""
        val matcher = matchPattern(activityPage, regex)
        val competitions = ArrayList<Int>()
        while (matcher.find()) {
            val divEndedClass = matcher.group(2)
            if ("contest_header ended" == divEndedClass) {
                competitions.add(matcher.group(1).toInt())
            }
        }
        return competitions
    }

    @Throws(IOException::class)
    fun getSecondsUntilNextCompetition(activityPage: String): Int {
        val regex = "data-remaining_time=\"([0-9]+)\""
        val matcher = matchPattern(activityPage, regex)
        var minTimeSeconds: Int? = null
        while (matcher.find()) {
            val waitTimeSeconds = matcher.group(1).toInt()
            if (minTimeSeconds == null || minTimeSeconds > waitTimeSeconds) {
                minTimeSeconds = waitTimeSeconds
            }
        }
        return minTimeSeconds ?: 24 * 60 * 60
    }

    @Throws(IOException::class)
    fun getChampionData(championContent: String): ChampionData? {
        val regex = "var championData = (\\{.*?});"
        val matcher = matchPattern(championContent, regex)
        return if (matcher.find()) {
            objectMapper.readValue<ChampionData>(matcher.group(1))
        } else null
    }

    fun getEventData(homeContent: String): EventData? {
        val regex = "event_data = (\\{.*});"
        val matcher = matchPattern(homeContent, regex, false)
        return if (matcher.find()) {
            objectMapper.readValue<EventData>(matcher.group(1))
        } else null
    }

    fun getChampionsIds(championsMapContent: String): List<ChampionPremise> {
        val regex = ("<a href=\"champions/([0-9]+)\" class=\"champion-lair\""
                + ".*?"
                + "<div class=\"champion-lair-name map-label-link\">"
                + "(.*?)"
                + "</div>")
        val matcher = matchPattern(championsMapContent, regex)
        val premises: MutableList<ChampionPremise> = ArrayList()
        while (matcher.find()) {
            val id = matcher.group(1).toInt()
            val waitTime = getWaitTime(matcher.group(2))
            premises.add(ChampionPremise(id, waitTime))
        }
        return premises
    }

    private fun getWaitTime(championLairContent: String): Int {
        val regex = "<div rel=\"timer\" timer=\"([0-9]+)\">"
        val matcher = matchPattern(championLairContent, regex)
        if (matcher.find()) {
            val timer = matcher.group(1).toLong()
            val waitTime = (timer - System.currentTimeMillis() / 1000).toInt()
            return if (waitTime < 0) 0 else waitTime
        }
        return 0
    }

    @Throws(IOException::class)
    fun getQuestSteps(questContent: String): Array<QuestStep> {
        val regex = "Q.steps = (\\[.*?\\]);"
        val matcher = matchPattern(questContent, regex)
        if (!matcher.find()) {
            return emptyArray()
        }
        val json = matcher.group(1).replace("\"cost\":[]", "\"cost\":{}")
        return objectMapper.readValue(json)
    }

    @Throws(IOException::class)
    fun findAllGirls(haremContent: String): List<Girl> {
        val girls: MutableList<Girl> = ArrayList()
        val regex = "girlsDataList\\['[0-9]+'\\] = (\\{.*?};)"
        val matcher = matchPattern(haremContent, regex)
        while (matcher.find()) {
            val girl = objectMapper.readValue<Girl>(matcher.group(1))
            girls.add(girl)
        }
        return girls
    }

    @Throws(IOException::class)
    fun getPlayerInfos(homeContent: String): UserInfos? {
        val regex = "Hero.infos = (\\{.*?};)"
        val matcher = matchPattern(homeContent, regex)
        return if (!matcher.find()) {
            null
        } else objectMapper.readValue<UserInfos>(matcher.group(1))
    }

    @Throws(IOException::class)
    fun findOpponentBattlePlayer(content: String): BattlePlayer? {
        val regex = "hh_battle_players =.*?\\{.*?},.*?(\\{.*?})"
        val matcher = matchPattern(content, regex)
        if (!matcher.find()) {
            return null
        }
        val battlePlayerStr = matcher.group(1)
        return objectMapper.readValue<BattlePlayer>(battlePlayerStr)
    }

    @Throws(IOException::class)
    fun findOpponentBattleMob(battleTrollContent: String): BattleMob? {
        val regex = "hh_battle_players =.*?\\{.*?},.*?(\\{.*?})"
        val matcher = matchPattern(battleTrollContent, regex)
        if (!matcher.find()) {
            return null
        }
        val battlePlayerStr = matcher.group(1)
        return objectMapper.readValue<BattleMob>(battlePlayerStr)
    }

    fun getCurrentWorldId(mapContent: String): String? {
        val regex = "<a class=\"link-world\" href=\"/world/([0-9]+)\""
        val matcher = matchPattern(mapContent, regex)
        var lastWorldId: String? = null
        while (matcher.find()) {
            lastWorldId = matcher.group(1)
        }
        return lastWorldId
    }

    fun getTrollId(worldContent: String): String? {
        val regex = "id_troll=([0-9]+)"
        val matcher = matchPattern(worldContent, regex)
        return if (matcher.find()) {
            matcher.group(1)
        } else null
    }

    fun getNextStock(shopContent: String): Int? {
        val regex = "<span rel=\"count\" time=\"([0-9]+)\">"
        val matcher = matchPattern(shopContent, regex)
        return if (matcher.find()) {
            matcher.group(1).toInt()
        } else null
    }

    @Throws(IOException::class)
    fun getShop(shopContent: String): Shop? {
        val regex = "<div tab class=\"armor\" type=\"armor\">(.*?)<button rel=\"buy\""
        val matcher = matchPattern(shopContent, regex)
        if (matcher.find()) {
            val subShopContent = matcher.group(1)
                    .replace("&quot;", "\"")
                    .replace("data-d=", "\ndata-d=")
            val books = getBooks(subShopContent)
            val gifts = getGifts(subShopContent)
            return Shop(books, gifts)
        }
        return null
    }

    @Throws(IOException::class)
    private fun getGifts(subShopContent: String): List<Item> {
        val regex = "data-d=\"(\\{.*?\"type\":\"gift\".*?})\">"
        val matcher = matchPattern(subShopContent, regex, false)
        val gifts: MutableList<Item> = ArrayList()
        while (matcher.find()) {
            gifts.add(objectMapper.readValue(matcher.group(1)))
        }
        return gifts
    }

    @Throws(IOException::class)
    private fun getBooks(subShopContent: String): List<Item> {
        val regex = "data-d=\"(\\{.*?\"type\":\"potion\".*?})\">"
        val matcher = matchPattern(subShopContent, regex, false)
        val books: MutableList<Item> = ArrayList()
        while (matcher.find()) {
            books.add(objectMapper.readValue(matcher.group(1)))
        }
        return books
    }

    @Throws(IOException::class)
    fun findHallOfFameOpponents(content: String): List<TowerOfFameOpponentPremise> {
        val regex = "leagues_list\\.push\\( (\\{.*?}) \\);"
        val matcher = matchPattern(content, regex)
        val premises: MutableList<TowerOfFameOpponentPremise> = ArrayList()
        while (matcher.find()) {
            premises.add(objectMapper.readValue(matcher.group(1)))
        }
        return premises
    }

    private fun matchPattern(content: String, regex: String, flatten: Boolean = true): Matcher {
        val treatedContent = if (flatten) flattenContent(content) else content
        val pattern = Pattern.compile(regex)
        return pattern.matcher(treatedContent)
    }

    private fun flattenContent(content: String): String {
        return content.replace("[\r\n]+".toRegex(), " ")
    }

    @Throws(IOException::class)
    fun getPlacesOfPower(activityPage: String): MutableList<PlaceOfPower> {
        val regex = "var pop_data.*?=.*?\\{.*?\\};"
        val matcher = matchPattern(activityPage, regex, false)
        if (matcher.find()) {
            val body = matcher.group(1)
            val popMap = objectMapper.readValue<Map<String, PlaceOfPower>>(body)
            return ArrayList(popMap.values)
        }
        return ArrayList()
    }

    @Throws(IOException::class)
    fun getMissions(activityPage: String): MutableList<Mission> {
        val missions: MutableList<Mission> = ArrayList()
        val regex = "<div class=\"mission_object mission_entry (.*?)\" data-d='(\\{\"id_member_mission\".*?})'"
        val matcher = matchPattern(activityPage, regex, false)
        while (matcher.find()) {
            val rarity = matcher.group(1)
            val missionBody = matcher.group(2)
            val mission = objectMapper.readValue<Mission>(missionBody)
            mission.setRarity(rarity)
            missions.add(mission)
        }
        return missions
    }

    fun getNextArenaReset(arenaContent: String): Long? {
        val regex = "\\.arena_refresh_counter \\[rel=\"count\"\\]'\\), ([0-9]+), reload\\);"
        val matcher = matchPattern(arenaContent, regex)
        if (matcher.find()) {
            return matcher.group(1).toLong()
        }
        return null
    }

    fun getEndOfWeekTimer(towerOfFameContent: String): Long? {
        val regex = "var till_end_of_week_timer = \\{\"in_seconds\":([0-9]+),\"days\":[0-9]+};"
        val matcher = matchPattern(towerOfFameContent, regex, false)
        return if (matcher.find()) {
            matcher.group(1).toLong()
        } else null
    }

    fun getEndOfLeagueTimer(towerOfFameContent: String): Long? {
        val regex = "var season_end_at = ([0-9]+);"
        val matcher = matchPattern(towerOfFameContent, regex, false)
        return if (matcher.find()) {
            matcher.group(1).toLong()
        } else null
    }

    fun getRewardsVar(towerOfFameContent: String): Int? {
        val regex = "var getRewards = ([0-9]+);"
        val matcher = matchPattern(towerOfFameContent, regex, false)
        return if (matcher.find()) {
            matcher.group(1).toInt()
        } else null
    }

    fun getNextMissionsUpdate(activityPage: String): Long? {
        val regex = "var next_update = ([0-9]+);"
        val matcher = matchPattern(activityPage, regex, false)
        return if (matcher.find()) {
            matcher.group(1).toLong()
        } else null
    }

    fun isLeagueRewardCollectible(towerOfFameContent: String): Boolean {
        val regex = "<div class=\"leagues_table lead_table\" >"
        val matcher = matchPattern(towerOfFameContent, regex, false)
        return matcher.find()
    }

}